### ISSUE TITLE
make raveberry kubernetes ready

### ISSUE THREAD

**Issue by skoef** [raveberry/raveberry#54]
I've been patching raveberry and the docker setup bit by bit to make it suitable to run in kubernetes. Some changes are kubernetes specific and some could be easily upstreamed anyway imho. Here is a list of things I changed to get it working in kubernetes:

- Since nginx serves the static files, there has to be a shared filesystem between the daphne and nginx containers. Obviously this could be dealt with with shared volumes across kubernetes pods, if that was available. To keep things simple, I added nginx to the daphne pod and created an entrypoint script for the daphne container to copy all static files to an emptyDir volume that was mounted on both the nginx and daphne container. This worked quite well, except for the compressed/cached javascript bundle in `CACHE/js/output.gibberish.js` which I couldn't directly figure out how to generate *before* copying the static files. I ended up enabling static file serving in django and removing the `location /static` part from the nginx config. This way I could separate the nginx container from the daphne container. This make scheduling and scaling both containers easier, if need might ever be. I kept the entrypoint file because I think this kind of logic (calling migrate and compiling the CSS) shouldn't be part of how you start the container.

- Using the youtube provider, *most of the time* the `EmbedThumbnail` postprocessor failed when I requested songs. I ended up disabling this and this works just fine. Not sure where these thumbnails would've been used anyway, but that said: I've only used the web interface so far.

- For using icecast, I removed the docker specific resolving part of the nginx config. I'm no nginx config expert, but perhaps by checking if nginx is running in kubernetes (for instance by checking of the `KUBERNETES_SERVICE_HOST` environment variable is set, you could have both solutions in one file for maintenance convenience.

- I'm in the process of creating a helm chart for easily deploying raveberry in kubernetes

Let me know if you want any of these changes backported to raveberry!

**Comment by raveberry** [comment]
Wow! First of all, I'm amazed you put in so much work. I have never worked with kubernetes (yet), so please excuse my ignorance.

- I guess that a shared filesystem is not the best design decision. However, I'm not entirely happy with daphne serving all static files either. The cleanest solution would probably be to include all static files in the nginx pod, as you mentioned. There is the `manage.py compress --force` command to create the javascript file beforehand. Then the static folder should actually remain static, making it possible to copy it over. This is a good idea regardless of kubernetes.
Also, you are right about the entrypoint. It is much better to not rely on the commands being given to the container.
If you want to create a pull request with these changes I will happily merge it, otherwise I will implement them sometime.

- Thumbnails are embedded because I might want to use them for some sort of display mode in the future. So no functionality breaks when you disable it. Unfortunately, Youtube recently changed the way they served thumbnails, so youtube-dl currently crashes when it tries to embed one. There already is a [pull request](https://github.com/ytdl-org/youtube-dl/pull/25717) for youtube-dl. I already disabled the embedding last week to deal with this problem.

- Some way to have the nginx.conf work in both setups would be nice, but not that important. There are already two different docker-compose.yml files, so it should not be too bad to add another version.

- As I mentioned, I have no experience with kubernetes. From what I understand them they are a sort of Dockerfile on a larger scale?

May I ask why you want to make raveberry run on kubernetes?

**Comment by skoef** [comment]
Awesome, I will send some PRs your way the following days!

Since you do appear to have experience with docker-compose: think of kubernetes as docker-compose on steroids. Instead of running docker-compose up and have all containers running locally or on one specific node, kubernetes allows and orchestrates running the containers on any node in a kubernetes cluster but remains the possibility to share network and volumes between containers. That, and a million other cool features ;)

**Comment by raveberry** [comment]
It's very nice that this project that started on a modest Raspberry Pi can soon be run with kubernetes :D
I'm not quite sure whether many people will use it, but it's definitely nifty.

**Comment by raveberry** [comment]
I just merged your pull request concerning the entrypoint. Also, I separated the daphne container from the nginx container by copying the static files into the nginx container. Now no shared volume is needed and static files don't need to be served by daphne.

**Comment by skoef** [comment]
I saw your updates, thanks! However, files like `static/scss/dark.css` are missing from the nginx container, right? Unless you've run the compilecss command locally before building the container? Same goes for all static admin files, those come from the django package and not this repo iirc.

**Comment by raveberry** [comment]
Yes, if the Dockerfile is built by itself, then these files are missing. The reason is that the django package and all of raveberry is needed to create these files. Since I wanted to include them in both the raveberry container (for the user-level server) and the nginx container (for static file serving in production), I decided to create them beforehand once.

You can take a look at https://github.com/raveberry/raveberry/blob/master/.travis.yml to see how it is done. Both containers are uploaded to [dockerhub](https://hub.docker.com/u/raveberry) after every update.

**Comment by skoef** [comment]
Another thing I noticed: when you update the spotify credentials, it runs `update_mopidy_config` which writes out the configuration locally, but when you run in a docker setup mopidy obviously runs in a different container so this doesn't work. I see that Mopidy supports a config API, would it be an option if raveberry pushes the configuration to Mopidy on startup and updates? That would work in both container and standalone configurations.

**Comment by raveberry** [comment]
Yes. Since they both run in different containers, raveberry can not overwrite the mopidy config file and restart the service. Thus, `update_mopidy_config` does not do anything in a docker setup. Users are required to provide the mopidy container with the correct config. I thought that this would not be too inconvenient, as it should change very rarely.

I guess you are talking about [this config api](https://docs.mopidy.com/en/latest/api/config/). I'm quite sure that this does not allow us to change the configuration of a running mopidy service. I understand it more as an internal api to parse and process config values. I'd be happy to be proven wrong, but I think that changing the config file and restarting the service is necessary to change mopidy's configuration.
Still, I don't think it's too much effort to manually edit the config file for the container. I know the credentials have to be passed to two containers currently, and it could be scripted, but since it's a one off thing I find it justifiable as it is.

**Comment by skoef** [comment]
I think this does the job then for now: https://github.com/raveberry/raveberry/pull/58

**Comment by raveberry** [comment]
Merged. What next? :D
